### PR TITLE
New 'appstudio-shared' module for AppStudio Environment APIs, add dependencies from appstudio-controller

### PR DESCRIPTION
**Pull request by jgwest**
#### Description:
- Introduces new `appstudio-shared` module, whose sole responsibility (at present) is to contain the Go structs and CRDs for the Environment API
    - Adding this as a separate module (rather than incorporating it into controller) allows us to eventually move this into its own Git repo and turn this into a module for containing other AppStudio APIs, in addition to ours (such as Application).
    - This separates the definition of the types and their behaviour.
- Adds a dependency to `appstudio-controller`, which will be responsible for handling the behaviour of these APIs.
- See [README.md](https://github.com/jgwest/managed-gitops/blob/env-api-may-2022/appstudio-shared/README.md) of appstudio-shared for initial introduction.
- API definition is based on [Environments API v3 Doc](https://docs.google.com/document/d/1-_rWLgALd5pdSlqNNcQ5FSrD00fZb0l_exU-_FiL68o/edit#)
- **Note**: This is only Go structs and CRDs, there is currently no implementation logic for any of these K8s resources.

#### Link to JIRA Story (if applicable): https://issues.redhat.com/browse/GITOPSRVCE-105

### CODE REVIEW

File: appstudio-shared/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+
+# AppStudio-Shared
+
+The APIs defined in this module are based on the [Environment API discussions](https://docs.google.com/document/d/1-_rWLgALd5pdSlqNNcQ5FSrD00fZb0l_exU-_FiL68o/) of April/May 2022.
```
Comment (samyak-jn) [inline appstudio-shared/README.md:4]:
```suggestion
The APIs defined in this module is based on the [Environment API discussions](https://docs.google.com/document/d/1-_rWLgALd5pdSlqNNcQ5FSrD00fZb0l_exU-_FiL68o/) of April/May 2022.
```

File: appstudio-shared/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+## Warning for API Consumers
+
+⚠️ Until these APIs are stabilized, they are subject to change with limited notice. Best efforts will me made to inform API consumers of how to migrate, when the changes are breaking. ⚠️
```
Comment (samyak-jn) [inline appstudio-shared/README.md:8]:
```suggestion
⚠️ Until these APIs are stabilized, they are subject to change with limited notice. Best efforts will be made to inform API consumers of how to migrate when the changes are breaking. ⚠️
```

File: appstudio-shared/README.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+### Expected upcoming changes, as of this writing (May 2022):
+- Drop the `Application-` prefix for prefixed resources
+- Move away from using `Snapshot` noun (replacement TBD)
```
Comment (samyak-jn) [inline appstudio-shared/README.md:12]:
```suggestion
- Move away from using the `Snapshot` noun (replacement TBD)
```

File: appstudio-shared/main.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	utilruntime.Must(appstudioredhatcomv1alpha1.AddToScheme(scheme))
+	//+kubebuilder:scaffold:scheme
+}
+
+func main() {
+	var metricsAddr string
+	var enableLeaderElection bool
+	var probeAddr string
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
+	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
```
Comment (samyak-jn) [inline appstudio-shared/main.go:55]: Just a newbie curiosity doubt, this means we are hardcoding the binding address right? Will that be the same for all, or do we plan to fetch the ports as well in the future?

File: .github/workflows/static.yml [hunk 5]
```diff
@@ -78,6 +78,11 @@ jobs:
           cd $GITHUB_WORKSPACE/appstudio-controller
           make lint 
 
+      - name: "Ensure 'appstudio-shared' passes 'go-lint' - run 'make lint' to identify issues"
+        run: |
+          cd $GITHUB_WORKSPACE/appstudio-shared
+          make lint 
+
```
Comment (samyak-jn) [inline .github/workflows/static.yml:85]: Do we also want to add a gosec check for this?

```
      - name: "Ensure 'appstudio-shared' passes 'go-sec' - run 'make gosec' to identify issues"
        run: |
          cd $GITHUB_WORKSPACE/appstudio-shared
          make gosec
```

**Review by samyak-jn — COMMENTED** [review]
Hi @jgwest :hugs:,

This is some awesome work there, In my review, the following are very nitty-gritty suggestions, otherwise, it's hard work on the functionality side (from v1->v3 of env API document :raised_hands:) Thanks!

LGTM :") :taco: :rocket:

**Review by jparsai — APPROVED** [review]
LGTM :)

**Review by drpaneas — APPROVED** [review]
lgtm

File: appstudio-shared/main.go (hunk 4 above)
Comment (jgwest) [reply appstudio-shared/main.go:55]: TLDR: this code doesn't actually run.

Long version: kubebuilder generates this main.go, and for some calls to operator-sdk, it requires this file to exist (AFAICT). Thus I'm leaving this file in, but disabling it from running. I've added additional statements to the file to clarify its purpose.

**Pull request merged**